src/shmem_atomic.h

```diff
@@ -57,72 +57,68 @@ static inline
 void
 shmem_spinlock_init(shmem_spinlock_t *lock)
 {
-    lock->enter = lock->exit = 0;
-    COMPILER_FENCE();
+    __atomic_store_n(&lock->exit, 0, __ATOMIC_RELEASE);
+    __atomic_store_n(&lock->enter, 0, __ATOMIC_RELEASE);
 }
 
 
 static inline
 void
 shmem_spinlock_lock(shmem_spinlock_t *lock)
 {
-    long val = __sync_fetch_and_add(&lock->enter, 1);
-    while (val != (lock->exit)) {
+    long val = __atomic_fetch_add(&lock->enter, 1, __ATOMIC_ACQUIRE);
+    while (val != __atomic_load_n(&lock->exit, __ATOMIC_RELAXED)) {
```

Comment:
Looking back at this ... the unlock operation performs a release on `lock->exit`.  To get ordering, I think we want the load of `lock->exit` here to be an acquire.  This eliminates the thread fence below.